Graph comparison update

matthid opened the pull request:
Discuss and merge.

matthid commented on codeface/R/cluster/graph_comparison.r:
```diff
@@ -91,8 +91,35 @@ vertex.neighborhood.difference <- function(g1, v1, g2, v2) {
+  }
+
+  return(difference)
+}
```
Because feature graph edges are always both directional I think vertices should be considered equal, when there is an corresponding incomming or outgoing edge.

mitchell-joblin replied:
@matthid I understand the rationale and it does make sense. I would instead modify the original function since it does share many commonalities. One possibility add another parameter to specify whether directionality should be considered and is the "as.undirected" function when necessary. Another possibility is to use the neigbors function with "mode=all" to achieve what you want.

matthid commented on codeface/R/cluster/graph_comparison.r:
```diff
@@ -142,37 +173,134 @@ graph.comparison <- function(g.1, g.2) {
+    map[[as.character(ids$id2[i])]] <- ids$id[i]
+  }
+  return(map)
+}
```
I think there is a nicer way to do this but my R is not good enough :)

mitchell-joblin replied:
@matthid Yeah R takes some getting used to, no worries. Perhaps you could pass in the column names as parameters instead of creating two functions. (see above)

matthid commented on codeface/R/cluster/graph_comparison.r:
```diff
@@ -142,37 +173,134 @@ graph.comparison <- function(g.1, g.2) {
   ##       to a clustering method
   cluster.method="Spin Glass Community"
+  pid.1 <- get.project.id.from.release.range.id(con, range.id.1)
+  pid.2 <- get.project.id.from.release.range.id(con, range.id.2)
```
I think range + projectId is unnecessary information? One can get the project from the range.

mitchell-joblin replied:
That is true but the goal in the end is probably to chose two project ids and compute the similarity over all ranges. You could just write a wrapper over this to do that though so I suppose it doesn't really matter.

matthid commented on codeface/R/cluster/graph_comparison.r:
```diff
@@ -142,37 +173,134 @@ graph.comparison <- function(g.1, g.2) {
-  edgelist.2   <- graph.data.2$edgelist
+  # get the raw edgelists with the ids from the database
+  edgelist.1   <- graph.data.1$edgelist.db
+  edgelist.2   <- graph.data.2$edgelist.db
```
`$edgelist` is some kind of magic edgelist, where the developer ids have been replaced with local ids.
But the local ids from edgelist.1 and edgelist.1 do not have to match because the list of developers could be different between two projects. Therefore we use the raw `$edgelist.db` here which has unmodified database developer ids.

mitchell-joblin replied:
@matthid I think it isn't necessary to do this. There is a mapping from local id to global id in the graph.data.* call id.map. The reason we create the edgelist with different ids thant the db global ids is because the one of the graph pacakges needs the vertex ids to indexed from 1. Maybe this is no longer required. Now we basically have get.graph.data.local returning two edgelists and if the graph is huge and sometimes we process them in parallel this is going to have an impact on memory.

matthid replied:
In this case it probably makes more sense to keep the edgelist.db (raw database ids) and remove the copied edgelist with magic local ids?

matthid commented on codeface/R/cluster/graph_comparison.r:
```diff
@@ -142,37 +173,134 @@ graph.comparison <- function(g.1, g.2) {
+  temp.merged <- merge (temp.vertex.df.1, temp.vertex.df.2, all=T, by="name")
+  merged.vertex <- data.frame(id=1:nrow(temp.merged), name=temp.merged$name, id1=temp.merged$id1, id2=temp.merged$id2)
+  map1 <- get.index.map1(merged.vertex)
+  map2 <- get.index.map2(merged.vertex)
```
Here I try to merge the two dataframes `temp.vertex.df.1` and `2` which contain the ids of the developers (same developer can have different ids) into a new dataframe `temp.merged`. And then I create local ids for the merged frame. From `merged.vertex` we can then create two mappings (one for `edgelist.1` and one for `edgelist.2`) to map the global developer ids into our new local id.

matthid commented on codeface/R/cluster/graph_comparison.r:
```diff
@@ -142,37 +173,134 @@ graph.comparison <- function(g.1, g.2) {
+  V(g.2)$Id <- merged.vertex$name
+
+  # vertex.diff describes how different the edges are, we now calculate how different the graphs are
+  vertex.diff <- graph.comparison(g.1, g.2, weighted = weighted, symmetric = symmetric)
```
I expect the comparison to return the same results for `vertices=merged.vertex` and `vertices=vertex.df.1` (but `merged.vertex` should return a larger vector (with more `1`s in it). But in my experiments this doesn't seem to be the case. So there is still something happening here which I do not understand...

matthid replied:
To clarify, I expected
```R
# A graph with weights and all vertices
g.1 <- graph.data.frame(edgelist.local.1, vertices=vertex.df.1, directed=directed)
g.2 <- graph.data.frame(edgelist.local.2, vertices=vertex.df.2, directed=directed)
V(g.1)$Id <- vertex.df.1$name
V(g.2)$Id <- vertex.df.2$name
vertex.diff <- graph.comparison(g.1, g.2, weighted = weighted, symmetric = symmetric)
```
to give the same `vertex.diff` vector (but with less ones) but that doesn't seem to be the case. I'll try to investigate this.

matthid replied:
I figured it out: When we use the limited vertex list the `graph.comparison` function will intersect the vertices and therefore "throw away" edges. Those edges are counted in the case with all vertices.

mitchell-joblin replied:
Yes exactly. The graph.comparison should also align the vertex ids based on the vertex name. It looks like you reimplemented that functionality.

matthid replied:
Not exactly: 
To compare the graphs you have two options: 
 * intersect the vertices (which was already implemented) 
 * or fill up the missing vertices (which I implemented).

They look similar but give different results. I think both are interesting.

matthid commented on codeface/R/cluster/graph_comparison.r:
```diff
@@ -142,37 +173,134 @@ graph.comparison <- function(g.1, g.2) {
+      total <- total + 1
+    }
+  }
+  vertex.total.diff <- total.diff / total
```
Some new metrices which (hopefully) make sense.

mitchell-joblin replied:
Looks interesting!

matthid commented on codeface/R/cluster/graph_comparison.r:
```diff
@@ -142,37 +173,134 @@ graph.comparison <- function(g.1, g.2) {
+               density.1 = graph.density(unmerged.unweighted.g.1),
+               density.2 = graph.density(unmerged.unweighted.g.2),
+               transitivity.1 = transitivity(unmerged.unweighted.g.1),
+               transitivity.2 = transitivity(unmerged.unweighted.g.2)))
```
Return all the data including some graph properties.

mitchell-joblin replied:
We have a lot of graph analysis code in another file but this is fine to give some context for now.

mitchell-joblin commented on codeface/R/cluster/graph_comparison.r:
```diff
@@ -142,37 +173,134 @@ graph.comparison <- function(g.1, g.2) {
   return(graph.diff)
 }
 
+get.index.map1 <- function(ids) {
```
get.index.map <- function(df, col1, col2) {
map <- vector(mode="list", nrow(df))
map[as.character(df[, col.1])] <- df[, col.2]
}

matthid commented on codeface/R/cluster/graph_comparison.r:
```diff
@@ -142,37 +151,248 @@ graph.comparison <- function(g.1, g.2) {
+  }
+  compare.ranges$id <- 1:len
+  merged <- merge(compare.ranges, overview, by="id", all=T)
+  return (list(overview=merged, vertexdata=vertexdata, graphs=graphdata))
```
This is the code I used for the data in the email, I figured it would be a nice addition.

matthid commented on codeface/R/cluster/graph_comparison.r:
```diff
@@ -125,14 +131,17 @@ graph.difference <- function(g1,g2, weighted=FALSE) {
 
-  intersectNames <- intersect(V(g.1)$Id, V(g.2)$Id)
-  idx.1 <- match(intersectNames, V(g.1)$Id)
-  idx.2 <- match(intersectNames, V(g.2)$Id)
```
You removed the `Id` column so I think your last fix (8ad7f04c28bdbe0b70a35d302dff96403bc3a692) will still fail here.

matthid commented:
I refactored the code a bit, extracted some functions and considered your last comments (thanks again!). I also had to rebase on top of your branch because of your latest changes (that's why the previous comments are now hidden).
